data/generate_embeddings.py

```diff
@@ -0,0 +1,598 @@
+    basename_counts: Dict[str, int] = {}
+    basename_temp: Dict[str, str] = {}
+
+    if not csv_path.exists():
+        print(f"Warning: metadata CSV not found at {csv_path}")
+        return path_to_city, {}
+
+    with csv_path.open("r", encoding="utf-8") as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            city = (row.get("city") or "").strip()
+            file_path = (row.get("file_path") or "").strip()
+            if not city or not file_path:
+                continue
+            rel_path = Path(file_path)
+            key_rel = rel_path.as_posix()
+            path_to_city[key_rel] = city
+            # Also map absolute path resolved from project root (cwd)
+            if not rel_path.is_absolute():
+                abs_key = (Path.cwd() / rel_path).resolve().as_posix()
+                path_to_city[abs_key] = city
+            base = rel_path.name
+            basename_counts[base] = basename_counts.get(base, 0) + 1
+            basename_temp[base] = city
+
+    basename_to_city: Dict[str, str] = {b: basename_temp[b] for b, c in basename_counts.items() if c == 1}
+    return path_to_city, basename_to_city
+
+
+def infer_city_from_metadata(input_path: Path, path_to_city: Dict[str, str], basename_to_city: Dict[str, str]) -> str:
+    """Infer city using metadata.csv mappings with multiple matching strategies."""
+    abs_posix = input_path.resolve().as_posix()
+    cwd_posix = Path.cwd().as_posix()
+    rel_from_cwd = abs_posix
+    if abs_posix.startswith(cwd_posix + "/"):
+        rel_from_cwd = abs_posix[len(cwd_posix) + 1 :]
+
+    # Try exact absolute path
+    city = path_to_city.get(abs_posix)
+    if city:
+        return city
+
+    # Try relative to project root
+    city = path_to_city.get(Path(rel_from_cwd).as_posix())
+    if city:
+        return city
+
+    # Fallback to basename if unique in metadata
+    base = input_path.name
+    city = basename_to_city.get(base)
+    if city:
+        return city
+
+    return "unknown"
+
+
+def ensure_output_dir(output_dir: Path):
+    """Create output directory if it doesn't exist."""
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+
+def get_next_global_chunk_id(output_dir: Path) -> int:
+    """Return the next global sequential chunk ID based on existing files.
+
+    Scans filenames like '<city>_NNN.json' and returns max(NNN)+1.
+    Note: This is O(n) over files in the output directory and is sufficient
+    for moderate corpus sizes. For very large directories, consider tracking
+    the last ID in a small sidecar file (e.g., 'index.json').
+    """
+    max_id = 0
+    for file_path in output_dir.glob("*.json"):
+        stem = file_path.stem
+        if "_" not in stem:
+            continue
+        id_part = stem.split("_")[-1]
+        if id_part.isdigit():
+            try:
+                max_id = max(max_id, int(id_part))
+            except ValueError:
+                continue
+    return max_id + 1
```

Comment:
If the output directory is very large, this could be slow; document the limitation or warn if file count is high.